Change to verification grid grayscale spectrograms

hudson-newey opened the pull request:
# Change to verification grid grayscale spectrograms

## Changes

- Bump web components version
- Use grayscale spectrograms in verification grid
- Add "loading" state
- Fix bug where clicking on context spectrogram (under "show more") would close the context popup and select the tile
- Migrate to esbuild application builder for tests
- Remove `@chiragrupani/karma-chromium-edge-launcher` as a dev dependency (we haven't tested on Edge in a long time)
- Enables linting for the `projects/` directory

### Bug Fixes

- Clicking on the "show more" button no longer selects the verification grid tile
- Clicking on the context spectrogram no longer selects the verification grid tile

## Issues

Fixes: #2152.

## Visual Changes

<img width="1025" height="843" alt="image" src="https://github.com/user-attachments/assets/660744bf-8971-40ea-a7d1-41244723471f" />

_Verification grid with grayscale spectrograms_

---

<img width="993" height="753" alt="image" src="https://github.com/user-attachments/assets/95ba159f-da17-4fff-8512-c43d7e3d4cd8" />

_grayscale spectrogram context_

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Change to verification grid grayscale spectrograms

This also fixes a bug where clicking on the context spectrogram would
close it.

Fixes: #2152

hudson-newey commented on src/app/components/web-components/grid-tile-content/grid-tile-content.component.ts:
```diff
@@ -43,23 +42,34 @@ export class GridTileContentComponent implements WithContext {
     }
 
-    const contextSize = 30 as const;
+    const contextSize = 30;
```
WHAT

atruskie replied:
😂

hudson-newey commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -466,7 +464,9 @@ class VerificationComponent
   private tagVerificationPredicate(): WhenPredicate {
     // The user can only verify a tag if there is a tag applied to the subject.
-    return (subject: SubjectWrapper) => subject.tag !== null;
+    return (subject: SubjectWrapper) => {
+      return subject.tag !== null;
+    }
```
I did this for improved debugging (e.g. adding `console.logs` returning dummy values, etc.

It's easier to debug because its easier to add lines.

This is a stylistic choice that I'm trying out recently.

```ts
const doSomething = () => {
	return 42;
};
```

than

```ts
const doSomething = () => 42;
```

atruskie commented on src/app/components/web-components/grid-tile-content/grid-tile-content.component.html:
```diff
@@ -11,7 +11,12 @@
+    We stop propagation here so that if you click anywhere in the context window
+    the click event won't bubble up to the grid-tile and cause the grid tile to
+    be selected/deselected.
+   -->
+  <div class="context-card context-anchor-item" (click)="$event.stopImmediatePropagation()">
     <div class="context-card-header">
       <span class="context-card-information">
         <span class="context-card-title">Context</span>
```
I thought this is fixed now with our click tracker patch?

hudson-newey replied:
Opening/closing the context window no longer selects the tile (because of our new click tracker patch).

However, because the spectrogram that is shown is technically slotted inside of the verification grid tile, clicking on the spectrogram would cause the verification grid tile to select.

By attaching a click event listener here, the click tracker patch stops the tile from being selected when you click the spectrogram.

hudson-newey committed:
Bump to 6.0.2

hudson-newey committed:
Align tests

hudson-newey committed:
Use application builder for tests

I did this because the webpack middleware was throwing errors.

Additionally, using webpack for tests while the real build uses esbuild
introduces bug risks.

I think Karma or esbuild might be doing something weird to the audio
files, because sometimes the test runner does a full reload (as if the
file changed), and the newly served file has different bytes from the
source file and fails to play in VLC or the spectrogram.
This isn't ideal, but at least the tests now run.

hudson-newey committed:
Should fix Windows CI

hudson-newey committed:
Only failing in Chrome

hudson-newey committed:
Trial verification grid with CDN

I'm pretty sure I've fixed all of the bugs sub 1.
If this passes CI, I know that I have 1 test fixture bug left, and it
also narrows it down to the Karma server.

hudson-newey committed:
Add comment explaining hacky base64

hudson-newey committed:
Bump for fixed worker destruction

hudson-newey committed:
Try disabling test

hudson-newey committed:
Staging to remove ng-mocks from metadata-review fixture

hudson-newey committed:
Mock providers with spectator

hudson-newey committed:
DONE

Hopefully this finishes this PR saga.

Metadata review tests are now passing.

The problem was that there were circular dependencies, that weren't
detected with webpack.

hudson-newey committed:
fix url directive tests

Webpack and esbuild URL constructors were behaving slightly different.
Webpack builds were returning document.hostURI as the test runners
document location, while esbuild was referring to the iframes uri.

This caused some problems with the expected urls in tests.

Note this was only a bug in the test fixture, not in the url directive.
The url directive has remained unchanged.

hudson-newey committed:
Try disabling verification tests to test timeout

CI is not showing what tests are timing out, but I suspect that it's the
verification grid tests.

Hopefully this should make CI fully pass,then I can incrementally
enabled focused verification tests.

hudson-newey committed:
Focus failing test

Is it really failing or just flakey?

hudson-newey committed:
Re-enable flakey test

hudson-newey committed:
Exclude test assets from linting & formatting

This should fix out of memory errors

hudson-newey committed:
Naieve

hudson-newey committed:
Focus verification tests to test flakyness

hudson-newey committed:
Found problematic tests

Tests that update the verification grids getPage callback before the
inital render cause tests to fail.

This still contains a fdescribe because I know that all of the other
tests should work, I just want to assert that cross-platform Chrome now
works as expected.

hudson-newey committed:
Change base64 to nominal typing

The base64 constant is quite large and because it's a string, it would
be typed as a structural type.

By changing it to a nominal type, we don't store the base64 audio file
in the type system.

hudson-newey committed:
Increase node memory

This is definately a hack, but I don't think the time investment into
finding the memory leak is worthwhile when we can just increase the node
memory (for now).

hudson-newey committed:
Unfocus verification tests

I think it should work now.

Docker is currently down, so I can't merge the PR.

hudson-newey committed:
Disable autowatch in CI

hudson-newey committed:
Use base64 in annotation event card tests

This was causing flakey tests on MacOS + Chrome

hudson-newey committed:
Merge branch 'master' into web-component-bump

hudson-newey committed:
use base64 on search page for less flakyness

hudson-newey committed:
Use devkit/build-angular instead of new angular/build

The new angular/build architect builder was resulting in "controller is
already closed" errors.
see: https://youtrack.jetbrains.com/issue/WEB-73511/Unable-to-run-tests-with-Karma-in-Angular-v20#focus=Comments-27-12564886.0-0

hudson-newey committed:
Revert "Use devkit/build-angular instead of new angular/build"

This reverts commit 4feebfb038544b76fb5ef7cff69b6e75b545b476.

hudson-newey committed:
Remove old devkit/build-angular from karma frameworks

hudson-newey committed:
Merge branch 'master' into web-component-bump

hudson-newey committed:
Significantly shorten base64 file

From 300kB to 20kB

hudson-newey committed:
Focused failing test with base64 url

hudson-newey committed:
Remove focused test
